src/mca/gds/base/gds_base_fns.c

```diff
@@ -168,6 +197,35 @@ pmix_status_t pmix_gds_base_store_modex(struct pmix_namespace_t *nspace,
                 goto error;
             }
 
+            /* calculate proc form the relative rank */
+            proc.rank = PMIX_RANK_UNDEF;
+            nspace_tmp = NULL;
+            PMIX_LIST_FOREACH(p, &plist_sort, pmix_proc_caddy_t) {
```

Comment:
Replace this O(N^2) algo with O(N)
1. Select unique namespaces from the proc list
2. Consider full namespace sizes instead of considering only ranks participating in the Fence. This will introduce gaps in the throughout numbering, but will not affect the correctness, but will improve packing and especially unpacking performance.